Readme.txt
Precedence is a raider's dashboard. It shows shot priority countdowns, warnings and timers
that are useful to a hunter or shadow priest who wants to maximize their DPS.

Supported classes:
* Hunters
* Shadow Priests

Shot/spell priorities:
* Configurable priorities
* Key bindings
* Use macros to trigger shots
* Time shots to buffs and debuffs
* Long cooldowns displayed in a useful way
* Delay all abilities during long casts & channeling

Warnings:
* Missing pet
* Incorrect aspects
* Missing hunter's mark
* Wrong weapon equipped
* Growl unexpectedly on/off
* Teleport ring/cloaks equipped

Timers:
* Misdirect applied, active & on cooldown (with target)
* Hunter's mark active
* Serpent sting active
* Mend pet active
* Freeze traps set & triggered (with target)



TODO
-----------------------------------------------------------------------------
* config for priorities

* Extra wait times for high priority shots
* Rules for serpent sting (don't use near end)
* Pet attacking?
* Deactivate/Fade out when dead
* Warn when wearing pvp gear
* Other trap timers
* Other stings

To bind a key, currently:

    /run PREC.options.priorities.p7.bind='ALT-7';

When switching between presets, update the options dialog (else opening and closing it seems to overwrite stuff?)


Changes.txt
1.0
 * first release

1.1
 * hide growl warnings when mounted
 * added warning for kirin tor ring
 * hide growl warnings when you don't have a pet
 * made the options pane scrollable
 * split config into 4 panes
 * added config for overall scale and runway length

1.2
 * updated for 4.x
 * set priority config from config screen
 * focus regen support
 * more shots
 * disable kill command when no pet is out
 * don't put up explosive shot when the debuff is on the target
 * warning tooltips

1.3
 * show bar correctly when mounted on a camel
 * show bad weapon warning for any weapons under ilvl 200
 * fixed boss & raid boss detection (hopefully)
 * more shots
 * correctly delay all shots while channeling steady & cobra shots
 * fixed wait time on explosive shot during L&L
 * added a bunch of MM specific abilities with built-in rules

1.4
 * updated for 4.2
 * no more pet happiness warning (option is still there, needs removing)

1.5
 * updated for 4.3
 * shadow priest support!
 * auto-hide for unsupported classes
 * hide during taxi rides
 * added arcane shot with reserved explosive

1.6
 * updated for 5.0.4
 * removed "sad pet" warning (deprecated in 4.2)
 * moved "bad weapon" warning from 'all' to 'priest'
 * added "non-ranged weapon" warning for hunters
 * added new hunter survival abilities (dire beast, crows)
 * added support for iron hawk aspect

1.7
 * added a new preset for 5.0.5 SV, including new cooldowns
 * added glaive toss and stampede
 * removed delay during L&L - explosive shot can be GCD chained now
 * fixed a bug with (raid)boss-only abilities appear as ready when they should be totally hidden
 * don't allow crows to come up as ready when it's already on the target (bug due to crows not having a timer)
 * don't allow readiness during rapid fire buff (bad time to waste a GCD)
 * new shadow priest rotation (with new orbs and mind spike buff)
 * removed vampiric embrace and shadow protection warnings
 * fixed 'bad weapon' warning for shadow priests to detect correct weapons
 * enforced recast delay no longer requires per-ability custom handlers
 * added enforced recast delay to mind spike
 * show bigger ability key labels for short labels/binds
 * don't let Dire Beast overlap after Readiness (probably good, hard to spend focus quick enough with 2 up)
 * include Dire Beast for normal mobs in in SV 5.0.5 rotation
 * for SV, allow Arcane Shot to wait for several higher-priority shots, not just Explosive
 * upgrade to interface 50100
 * fix for older saved vars - don't die if we have a prio set to an ability that does not exist
 * updated priest rotation for 5.1


class_all.lua
PREC.default_options.warnings.kirin_tor_ring = false;
PREC.default_options.warnings.guild_cloak = false;

PREC.warningdefs.kirin_tor_ring = {
	title = "Kirin Tor Ring Equipped",
	icon = [[Interface\Icons\inv_jewelry_ring_74]],
};

PREC.warningdefs.guild_cloak = {
	title = "Guild Cloak Equipped",
	icon = [[Interface\Icons\inv_guild_cloak_alliance_a]],
};

function PREC.warningdefs.kirin_tor_ring.func(info)

	local itemId1 = GetInventoryItemID("player", 11);
	local itemId2 = GetInventoryItemID("player", 12);

	if (itemId1 == 51560 or itemId2 == 51560) then info.show = true; end
	if (itemId1 == 48954 or itemId2 == 48954) then info.show = true; end
	if (itemId1 == 45688 or itemId2 == 45688) then info.show = true; end
	if (itemId1 == 40586 or itemId2 == 40586) then info.show = true; end
end

function PREC.warningdefs.guild_cloak.func(info)

	local itemId = GetInventoryItemID("player", 15);

	if (itemId == 63352) then info.show = true; end -- alliance 8hr
	if (itemId == 63353) then info.show = true; end -- horde 8hr
	if (itemId == 63206) then info.show = true; end -- alliance 4hr
	if (itemId == 63207) then info.show = true; end -- horde 4hr
	if (itemId == 65360) then info.show = true; end -- alliance 2hr
	if (itemId == 65274) then info.show = true; end -- horde 2hr
end


class_priest.lua
if select(2, UnitClass('player')) ~= "PRIEST" then return end

PREC.enable_for_class = true;

PREC.default_options.meters = {

	sw_pain = true,
	vampiric_touch = true,
	devouring_plague = true,
};

PREC.default_options.warnings = {

	fort = true,
	inner_fire = true,
	shadowform = true,
	bad_weapon = true, -- fishing pole, lance
};

PREC.rotations = {
	sp505 = {
		name = "Shadow Priest (5.0.5)",
		p1 = { which="sw_pain",		bind="",	who="any", label="1" },
		p2 = { which="vampiric_touch",	bind="",	who="any", label="3" },
		p3 = { which="plague_three",	bind="",	who="any", label="4" },
		p4 = { which="mind_blast",	bind="",	who="any", label="2" },
		p5 = { which="surge_spike",	bind="",	who="any", label="6" },
		p6 = { which="sw_death",	bind="",	who="any", label="5" },
		p7 = { which="mind_flay",	bind="",	who="any", label="8" },
		p8 = {},
		p9 = {},
		p10 = {},
		p11 = {},
		p12 = {},
	},
	sp510 = {
		name = "Shadow Priest (5.1.0)",
		p1 = { which="plague_three",	bind="",	who="any", label="4" },
		p2 = { which="mind_blast",	bind="",	who="any", label="2" },
		p3 = { which="sw_death",	bind="",	who="any", label="5" },
		p4 = { which="halo",		bind="",	who="any", label="9" },
		p5 = { which="surge_spike",	bind="",	who="any", label="6" },
		p6 = { which="vampiric_touch",	bind="",	who="any", label="3" },
		p7 = { which="sw_pain",		bind="",	who="any", label="1" },
		p8 = { which="mind_flay",	bind="",	who="any", label="8" },
		p9 = {},
		p10 = {},
		p11 = {},
		p12 = {},
	},
};

PREC.default_options.priorities = PREC.rotations.sp505;

PREC.abilities = {
	sw_pain = {
		icon = "spell_shadow_shadowwordpain",
		spell = "Shadow Word: Pain",
		debuff = "Shadow Word: Pain",
		delay_recast_for = 1,
	},
	mind_blast = {
		icon = "spell_shadow_unholyfrenzy",
		spell = "Mind Blast",
		delay_recast_for = 4,
	},
	vampiric_touch = {
		icon = "spell_holy_stoicism",
		spell = "Vampiric Touch",
		debuff = "Vampiric Touch",
		delay_recast_for = 1,
	},
	devouring_plague = {
		icon = "spell_shadow_devouringplague",
		spell = "Devouring Plague",
	},
	sw_death = {
		icon = "spell_shadow_demonicfortitude",
		spell = "Shadow Word: Death",
		label = "Shadow Word: Death (under 25%)",
	},
	archangel = {
		icon = "ability_priest_archangel",
		spell = "Archangel",
		havebuff = "Dark Evangelism",
		buff_stacks = 5,
		label = "Archangel (5 stacks)",
	},
	archangel_any = {
		icon = "ability_priest_archangel",
		spell = "Archangel",
		label = "Archangel (1 stack)",
	},
	shadowfiend = {
		icon = "spell_shadow_shadowfiend",
		spell = "Shadowfiend",
		-- only at 80% mana
	},
	mind_flay = {
		icon = "spell_shadow_siphonmana",
		spell = "Mind Flay",
	},
	plague_three = {
		icon = "spell_shadow_devouringplague",
		spell = "Devouring Plague",
		label = "Devouring Plague (w/ 3 Orbs)",
	},
	surge_spike = {
		icon = "spell_priest_mindspike",
		spell = "Mind Spike",
		havebuff = "Surge of Darkness",
	},
	halo = {
		icon = "ability_priest_halo_shadow",
		spell = "Halo",
	},
};

PREC.meterinfo = {
	sw_pain = {
		icon = "spell_shadow_shadowwordpain",
		debuff = "Shadow Word: Pain",
		color = "green",
	},
	vampiric_touch = {
		icon = "spell_holy_stoicism",
		debuff = "Vampiric Touch",
		color = "green",
	},
	devouring_plague = {
		icon = "spell_shadow_devouringplague",
		debuff = "Devouring Plague",
		color = "green",
	},
}

PREC.warningdefs = {
	fort = {
		title = "Buffed: Fortitude",
		icon = [[Interface\Icons\spell_holy_wordfortitude]],
		has_buff = "Power Word: Fortitude",
	},
	inner_fire = {
		title = "Buffed: Inner Fire",
		icon = [[Interface\Icons\spell_holy_innerfire]],
		has_buff = "Inner Fire",
	},
	shadowform = {
		title = "Shadowform",
		icon = [[Interface\Icons\spell_shadow_shadowform]],
		has_buff = "Shadowform",
		scale = 2,
	},
	bad_weapon = {
		title = "Bad Weapon Equipped",
		icon = [[Interface\Icons\inv_weapon_shortblade_05]],
		--not_implemented = true,
	}
};

PREC.state.simulate_focus_loss_until = 0;
PREC.state.no_shots_until = 0;
PREC.state.delay_spell_until = {};

function PREC.OnCombatLog(...)

	local ts, event, hideCaster, sourceGuid, sourceName, sourceFlags, sourceFlags2, destGuid, destName, destFlags, destFlasg2, spellId, spellName, spellSchool = ...;

	local srcUs = false;
	local ourGuid = UnitGUID("player");
	local tarGuid = UnitGUID("target");
	if (sourceGuid == ourGuid) then srcUs = true; end

	if (false and (srcUs or sourceGuid == tarGuid)) then
		local dest = destName;
		if (not dest) then dest = "?"; end
		local src = sourceName;
		if (not src) then src = "?"; end
		local spell = spellName;
		if (not spell) then spell = "?"; end

		print(string.format("%s (%s -> %s) %s", event, src, dest, spell));
	end
end

function PREC.OnSpellCastSent(...)

	local unit, spell, rank, target = ...;


	--
	-- spells with cast time need to delay all others
	--

	if (unit == "player") then
		local _, _, _, _, _, _, castTime = GetSpellInfo(spell);

		if (castTime) then

			PREC.state.no_shots_until = GetTime() + (castTime / 1000);

			-- supress this spell for cast time + cooldown, since
			-- we know it has a cooldown and we're casting it now.
			-- we do the cooldown calc in the update function since
			-- we can't get cooldown until cast has succeeded
			PREC.state.delay_spell_until[spell] = PREC.state.no_shots_until;
		end
	end
end


--
-- special event handlers for updating channeled spell status
--

function PREC.event_handlers.UNIT_SPELLCAST_CHANNEL_START(unit, spellName, spellRank, lineID, spellID)

	local _,_,_,_,_,endTime = UnitChannelInfo("player");
	if (endTime) then
		PREC.state.no_shots_until = endTime / 1000;
	else
		PREC.state.no_shots_until = 0;
	end
end

function PREC.event_handlers.UNIT_SPELLCAST_CHANNEL_UPDATE(unit, spellName, spellRank, lineID, spellID)

	local _,_,_,_,_,endTime = UnitChannelInfo("player");
	if (endTime) then
		PREC.state.no_shots_until = endTime / 1000;
	else
		PREC.state.no_shots_until = 0;
	end
end

function PREC.event_handlers.UNIT_SPELLCAST_CHANNEL_STOP(unit, spellName, spellRank, lineID, spellID)

	PREC.state.no_shots_until = 0;
end


function PREC.GetWarningLabel(ret)

	ret.show_label = false;

	--
	-- range warning
	--

	local inShotRange = IsSpellInRange("Mind Sear");

	if ((ret.active_shots > 0) and ret.has_viable_target) then

		if (not (inShotRange == 1)) then

			ret.show_label = true;
			ret.label = 'Too Far';
			ret.label_mode = 'warning';
			return;
		end
	end


	--
	-- mana warning
	--

	local cur_mana = UnitPower("player", 0);
	local max_mana = UnitPowerMax("player", 0);
	local mana_percent = 100 * cur_mana / max_mana;

	if (mana_percent < 10) then

		ret.show_label = true;
		ret.label = string.format('Mana Low (%d%%)', mana_percent);
		ret.label_mode = 'mana';
		return;
	end

end

--
-- ################################################## Warnings ##################################################
--

--function PREC.warningdefs.no_pet.func(info)
--end

function PREC.warningdefs.bad_weapon.func(info)

	local itemId = GetInventoryItemID("player", 16);

	if (not itemId) then return info; end

	local _, _, _, level, _, type, subtype = GetItemInfo(itemId);

	if (level) then
		if (level < 200) then info.show = true; end
	end

	if (type == "Weapon" and subtype == "Daggers"		) then return info; end
	if (type == "Weapon" and subtype == "One-Handed Maces"	) then return info; end
	if (type == "Weapon" and subtype == "Staves"		) then return info; end
	if (type == "Weapon" and subtype == "Wands"		) then return info; end

	info.show = true;
end

--
-- ################################################## Meters ##################################################
--

--function PREC.meterinfo.trap_set.func(info)
--end


--
-- ################################################## Shots ##################################################
--

function PREC.abilities.sw_death.func(t, now, waitmana)

	local current = UnitHealth("target");
	local max = UnitHealthMax("target");

	-- no target?
	if (max == 0) then
		return {
			hide_now = true,
		};
	end

	-- over 25%?
	if (current / max > 0.25) then
		return {
			hide_now = true,
		};
	end

	return {
		t = t,
		waitmana = waitmana,
	};
end

function PREC.abilities.plague_three.func(t, now, waitmana)

	local orbs = UnitPower("player", 13);
	if (orbs < 3) then
		return {
			hide_now = true,
		};
	end

	return {
		t = t,
		waitmana = waitmana,
	};
end


class_hunter.lua
if select(2, UnitClass('player')) ~= "HUNTER" then return end

PREC.enable_for_class = true;

PREC.default_options.chimera_refresh_window = 4;
PREC.default_options.viper_mana_bigger = 90;
PREC.default_options.mana_low_warning = 10;

PREC.default_options.meters = {
	md_applied = true,
	md_cooldown = true,
	hunters_mark = true,
	serpent_sting = false,
	mend_pet = true,
	trap_set = true,
	trap_triggered = true,
};

PREC.default_options.warnings = {
	no_pet = true,
	bad_aspect = true,
	no_hunters_mark = true,
	hunter_weapon = true,
	low_ammo = true,
	growl_solo = true,
	growl_party = true,
	kirin_tor_ring = false,
	guild_cloak = false,	
};

PREC.rotations = {
	sv406 = {
		name = "SV 4.0.6",
		p1  = { which="rapid",		bind="ALT-1",	who = "boss" },
		p2  = { which="kill",		bind="ALT-2",	who = "any" },
		p3  = { which="explosive",	bind="ALT-3",	who = "any" },
		p4  = { which="black",		bind="ALT-4",	who = "any" }, --label = "Chim",
		p5  = { which="serpent",	bind="ALT-5",	who = "any" },
		p6  = { which="arcane_es",	bind="ALT-7",	who = "any" },
		p7  = { which="cobra",		bind="ALT-6",	who = "any" }, --cmd = "MACRO Steady",
		p8  = {},
		p9  = {},
		p10 = {},
		p11 = {},
		p12 = {},
	},
	mm406 = {
		name = "MM 4.0.6",
		p1 = { which="serpent",		bind="ALT-1",	who="any" },
		p2 = { which="chimerarefresh",	bind="ALT-2",	who="any" },
		p3 = { which="aimedmmm",	bind="ALT-3",	who="any" },
		p4 = { which="steadynoiss",	bind="ALT-4",	who="any" },
		p5 = { which="aimed",		bind="ALT-3",	who="any" },
		p6 = { which="kill",		bind="ALT-5",	who="any" },
		p7 = { which="rapid",		bind="ALT-6",	who="boss" },
		p8 = { which="readiness",	bind="ALT-7",	who="boss" },
		p9 = { which="steady",		bind="ALT-4",	who="any" },
		p10 = {},
		p11 = {},
		p12 = {},
	},
	{
		name = "SV 5.0.5",
		p1  = { which="black",		bind="ALT-4",	who="any" },
		p2  = { which="crows",		bind="",	who="boss",	label="7" },
		p3  = { which="dire_beast",	bind="",	who="any",	label="8" },
		p4  = { which="stampede",	bind="ALT-8",	who="boss" },
		p5  = { which="rapid",		bind="ALT-1",	who="boss" },
		p6  = { which="glaive_toss",	bind="",	who="any",	label="2" },
		p7  = { which="kill",		bind="ALT-2",	who="any" },
		p8  = { which="explosive",	bind="ALT-3",	who="any" },
		p9  = { which="readiness",	bind="",	who="boss",	label="5" },
		p10 = { which="serpent",	bind="ALT-5",	who="any" },
		p11 = { which="arcane_sv",	bind="ALT-7",	who="any" },
		p12 = { which="cobra",		bind="ALT-6",	who="any" },
	},
};

PREC.default_options.priorities = PREC.rotations.sv406;

PREC.abilities = {
	rapid = {
		icon = [[ability_hunter_runningshot]],
		spell = "Rapid Fire",
		buff = "Rapid Fire",
	},
	kill = {
		icon = [[ability_hunter_assassinate2]],
		spell = "Kill Shot",
	},
	serpent = {
		icon = [[ability_hunter_quickshot]],
		spell = "Serpent Sting",
		debuff = "Serpent Sting",
	},
	chimera = {
		icon = [[ability_hunter_chimerashot2]],
		spell = "Chimera Shot",
	},
	chimerarefresh = {
		icon = [[ability_hunter_chimerashot2]],
		spell = "Chimera Shot",
		chimerarefresh = true,
	},
	aimedmmm = {
		icon = [[inv_spear_07]],
		spell = "Aimed Shot",
		havebuff = "Fire!",
		label = "Aimed Shot (MMM Proc)",
	},
	aimed = {
		icon = [[inv_spear_07]],
		spell = "Aimed Shot",
	},
	trapfrost = {
		icon = [[spell_frost_freezingbreath.jpg]],
		spell = "Frost Trap",
	},
	readiness = {
		icon = "ability_hunter_readiness",
		spell = "Readiness",
		buff = "Rapid Fire",
	},
	steady = {
		icon = "ability_hunter_steadyshot",
		spell = "Steady Shot",
	},
	steadynoiss = {
		icon = "ability_hunter_steadyshot",
		spell = "Steady Shot",
		gainiss = true,
		label = "Steady Shot (No ISS)",
	},
	explosive = {
		icon = "ability_hunter_explosiveshot",
		spell = "Explosive Shot",
		--debuff = "Explosive Shot",
	},
	black = {
		icon = "spell_shadow_painspike",
		spell = "Black Arrow",
		debuff = "Black Arrow",
	},
	killcmd = {
		icon = [[ability_hunter_killcommand]],
		spell = "Kill Command",
	},
	cobra = {
		icon = "ability_hunter_cobrashot",
		spell = "Cobra Shot",
	},
	arcane = {
		icon = "ability_impalingbolt",
		spell = "Arcane Shot",
	},
	arcane_es = {
		icon = "ability_impalingbolt",
		spell = "Arcane Shot",
		label = "Arcane Shot (reserve Explosive)",
		between_es = true,
	},
	arcane_sv = {
		icon = "ability_impalingbolt",
		spell = "Arcane Shot",
		label = "Arcane Shot (reserve SV)",
	},
	dire_beast = {
		icon = "ability_hunter_sickem",
		spell = "Dire Beast",
	},
	crows = {
		icon = "ability_hunter_murderofcrows",
		spell = "A Murder of Crows",
		debuff = "A Murder of Crows",
	},
	stampede = {
		icon = "ability_hunter_bestialdiscipline",
		spell = "Stampede",
	},
	glaive_toss = {
		icon = "ability_glaivetoss",
		spell = "Glaive Toss",
	},
	
};

PREC.meterinfo = {
	md_applied = {
		title = "Misdirect Active",
		icon = "ability_hunter_misdirection",
		buff = "Misdirection",
		color = "green",
		special_label = "md_target",
	},
	md_cooldown = {
		title = "Misdirect Cooldown",
		icon = "ability_hunter_misdirection",
		spell = "Misdirection",
		color = "red",
		label = "Cooldown",
	},
	hunters_mark = {
		icon = "ability_hunter_snipershot",
		debuff = "Hunter's Mark",
		color = "green",
	},
	serpent_sting = {
		icon = "ability_hunter_quickshot",
		debuff = "Serpent Sting",
		color = "green",
	},
	mend_pet = {
		icon = "ability_hunter_mendpet",
		petbuff = "Mend Pet",
		color = "green",
	},
	trap_set = {
		title = "Freeze Trap Set",
		color = "green",
	},
	trap_triggered = {
		title = "Freeze Trap Triggered",
		color = "green",
	},
}

PREC.warningdefs = {
	no_pet = {
		title = "Missing Pet",
		icon = [[Interface\Icons\inv_box_petcarrier_01]],
	},
	bad_aspect = {
		title = "Wrong Aspect",
		icon = [[Interface\Icons\spell_nature_ravenform]],
	},
	no_hunters_mark = {
		title = "Missing Hunter's Mark",
		icon = [[Interface\Icons\ability_hunter_snipershot]],
	},
	growl_solo = {
		title = "Growl Disabled When Solo",
		icon = [[Interface\Icons\ability_physical_taunt]],
	},
	growl_party = {
		title = "Growl Enabled In Party",
		icon = [[Interface\Icons\ability_physical_taunt]],
	},
	hunter_weapon = {
		title = "Non-Ranged Weapon Equipped",
		icon = [[Interface\Icons\inv_misc_ammo_arrow_04]],
	},
};

PREC.state = {
	md_target = "?",
	trap_set = false,
	trap_set_start = 0,
	trapped_mobs = {},
	no_shots_until = 0,
	no_explosive_until = 0,
	no_dire_beast_until = 0,
	simulate_focus_loss = 0,
	simulate_focus_loss_until = 0,
	steady_shots_accum = 0,
};

function PREC.OnCombatLog(...)

	local ts, event, hideCaster, sourceGuid, sourceName, sourceFlags, sourceFlags2, destGuid, destName, destFlags, destFlasg2, spellId, spellName, spellSchool = ...;

	local srcUs = false;
	local ourGuid = UnitGUID("player");
	if (sourceGuid == ourGuid) then srcUs = true; end

	if (srcUs and event == "SPELL_CAST_SUCCESS" and spellName == "Misdirection") then
		PREC.state.md_target = destName;
	end

	if ((event == "SPELL_CREATE") and (srcUs) and ((spellName == "Freezing Arrow") or (spellName == "Freezing Trap"))) then
		PREC.state.trap_set = true;
		PREC.state.trap_set_start = GetTime();
		if (spellName == "Freezing Trap") then PREC.meterinfo.trap_set.icon = "spell_frost_chainsofice"; end
		if (spellName == "Freezing Arrow") then PREC.meterinfo.trap_set.icon = "spell_frost_chillingbolt"; end
		return;
	end

	if ((event == "SPELL_MISSED") and srcUs and ((spellName == "Freezing Arrow Effect") or (spellName == "Freezing Trap Effect"))) then
		PREC.state.trap_set = false;
	end

	if ((event == "SPELL_AURA_APPLIED") and srcUs and ((spellName == "Freezing Arrow Effect") or (spellName == "Freezing Trap Effect"))) then

		PREC.state.trap_set = false;
		PREC.state.trapped_mobs[destGuid] = {
			start = GetTime(),
			aura = spellName,
			guid = destGuid,
			name = destName,
		};
		return;
	end

	if ((event == "SPELL_AURA_REMOVED") and srcUs and ((spellName == "Freezing Arrow Effect") or (spellName == "Freezing Trap Effect"))) then

		PREC.state.trapped_mobs[destGuid] = nil;
		return;
	end

	if (false) then
		local dest = destName;
		if (not dest) then dest = "?"; end
		local src = sourceName;
		if (not src) then src = "?"; end
		local spell = spellName;
		if (not spell) then spell = "?"; end

		print(string.format("%s (%s -> %s) %s", event, src, dest, spell));
	end
end

function PREC.OnSpellCastSent(...)

	local unit, spell, rank, target = ...;

	if (unit == "player") then
		if (spell == "Steady Shot") then
			PREC.state.steady_shots_accum = PREC.state.steady_shots_accum + 1;
		else
			PREC.state.steady_shots_accum = 0;
		end
	end

	if (unit == "player" and spell == "Cobra Shot") then
		local _, _, _, _, _, _, castTime = GetSpellInfo("Cobra Shot")
		PREC.state.no_shots_until = GetTime() + (castTime / 1000);
		return;
	end

	if (unit == "player" and spell == "Steady Shot") then
		local _, _, _, _, _, _, castTime = GetSpellInfo("Steady Shot")
		PREC.state.no_shots_until = GetTime() + (castTime / 1000);
		return;
	end

	if (unit == "player" and spell == "Aimed Shot") then
		local test = PREC.TimeToPlayerBuffExpires("Fire!", false);
		if (test == -1) then
			local _, _, _, cost, _, _, castTime = GetSpellInfo("Aimed Shot")

			local cast_ends = GetTime() + (castTime / 1000);

			PREC.state.no_shots_until = cast_ends;
			PREC.state.simulate_focus_loss_until = cast_ends;
			PREC.state.simulate_focus_loss = cost;
		end
		return;
	end

	if (unit == "player" and spell == "Explosive Shot") then
		--PREC.state.no_explosive_until = GetTime() + 2;
		return;
	end

	if (unit == "player" and spell == "Dire Beast") then
		PREC.state.no_dire_beast_until = GetTime() + 15;
		return;
	end
end

function PREC.GetWarningLabel(ret)

	ret.show_label = false;


	--
	-- range warning
	--

	local inShotRange = IsSpellInRange("Auto Shot");
	local inMeleeRange = IsSpellInRange("Wing Clip");

	if ((ret.active_shots > 0) and ret.has_viable_target) then

		if (not (inShotRange == 1)) then
			if (inMeleeRange == 1) then

				ret.show_label = true;
				ret.label = 'Too Close';
				ret.label_mode = 'warning';
				return;
			else
				ret.show_label = true;
				ret.label = 'Too Far';
				ret.label_mode = 'warning';
				return;
			end
		end
	end


	--
	-- focus warning
	--

	local cur_focus = UnitPower("player", 2);
	local max_focus = UnitPowerMax("player", 2);
	local focus_percent = 100 * cur_focus / max_focus;

	if (focus_percent < 10) then

		ret.show_label = true;
		ret.label = string.format('Low Focus (%d%%)', focus_percent);
		ret.label_mode = 'mana';
		return;
	end

	if (focus_percent > 80 and focus_percent < 100) then

		ret.show_label = true;
		ret.label = 'High Focus';
		ret.label_mode = 'mana';
		return;
	end
end

--
-- ################################################## Warnings ##################################################
--

function PREC.warningdefs.no_pet.func(info)

	if (IsMounted()) then return info; end
	if (UnitGUID("pet")) then
		if (UnitHealth("pet") == 0) then
			info.show = true;
			return info;
		end
		return info;
	end
	info.show = true;
	return info;
end

function PREC.warningdefs.bad_aspect.func(info)

	local bad_icon = nil;
	local found_hawk = false;

	local index = 1
	while UnitBuff("player", index) do
		local name, _, _, count, _, _, buffExpires, caster = UnitBuff("player", index)
		if (name == "Aspect of the Cheetah"	) then bad_icon = "ability_mount_jungletiger"; end
		if (name == "Aspect of the Fox"		) then bad_icon = "ability_hunter_aspectofthefox"; end
		if (name == "Aspect of the Pack"	) then bad_icon = "ability_mount_whitetiger"; end
		if (name == "Aspect of the Wild"	) then bad_icon = "spell_nature_protectionformnature"; end
		if (name == "Aspect of the Hawk"	) then found_hawk = true; end
		if (name == "Aspect of the Iron Hawk"	) then found_hawk = true; end
		index = index + 1
	end

	if (bad_icon) then
		info.show = true;
		info.icon = [[Interface\Icons\]] .. bad_icon;
		info.scale = 1;
	else
		if (not found_hawk) then
			info.show = true;
			info.icon = [[Interface\Icons\spell_nature_ravenform]];
		end
	end
	return info;
end

function PREC.warningdefs.no_hunters_mark.func(info)

	if (PREC.HasViableTarget()) then

		local temp = PREC.CheckBuff(UnitDebuff, "Hunter's Mark", "target", false);
		if (temp.t == 0) then
			info.show = true;
		end
	end
end

function PREC.warningdefs.growl_solo.func(info)

	if (IsMounted()) then return info; end
	if (not UnitGUID("pet")) then return info; end
	if (UnitHealth("pet") == 0) then return info; end

	local _, autostate = GetSpellAutocast("Growl", "pet");
	if (not autostate and not PREC.InGroup()) then
		info.show = true;
	end
end

function PREC.warningdefs.growl_party.func(info)

	if (IsMounted()) then return info; end
	if (not UnitGUID("pet")) then return info; end
	if (UnitHealth("pet") == 0) then return info; end

	local _, autostate = GetSpellAutocast("Growl", "pet");
	if (autostate and PREC.InGroup()) then
		info.show = true;
	end
end

function PREC.warningdefs.hunter_weapon.func(info)

	local itemId = GetInventoryItemID("player", 16);

	if (not itemId) then return info; end

	local _, _, _, level, _, type, subtype = GetItemInfo(itemId);

	if (level) then
		if (level < 200) then info.show = true; end
	end

	if (type == "Weapon" and subtype == "Bows"		) then return info; end
	if (type == "Weapon" and subtype == "Crossbows"		) then return info; end
	if (type == "Weapon" and subtype == "Guns"		) then return info; end

	info.show = true;
end

--
-- ################################################## Meters ##################################################
--

function PREC.meterinfo.trap_set.func(info)

	if (PREC.state.trap_set) then

		local duration = GetTime() - PREC.state.trap_set_start;
		local max = 30;

		if (duration > max) then
			PREC.state.trap_set = false;
		else
			info.max = max;
			info.t = max - duration;
		end
	end
end

function PREC.meterinfo.trap_triggered.func(info)

	info.multi = {};

	local guid, details;
	for guid, details in pairs(PREC.state.trapped_mobs) do

		local info2 = PREC.CopyTable(info);
		local duration = GetTime() - details.start;
		local max = 20;

		if (duration > max) then
			info2.t = 0.1;
			info2.max = max;
		else
			info2.max = max;
			info2.t = max - duration;
		end		

		info2.label = details.name;

		if (details.aura == "Freezing Trap Effect") then info2.icon = "spell_frost_chainsofice"; end
		if (details.aura == "Freezing Arrow Effect") then info2.icon = "spell_frost_chillingbolt"; end

		info.multi[guid] = info2;
	end
end

--
-- ################################################## Shots ##################################################
--

function PREC.abilities.explosive.func(t, now, waitmana)

	if (PREC.state.no_explosive_until > now) then
		local ex_min = PREC.state.no_explosive_until - now;
		if (t < ex_min) then
			t = ex_min;
		end		
	end

	return {
		t = t,
		waitmana = waitmana,
	};
end

function PREC.abilities.dire_beast.func(t, now, waitmana)

	if (PREC.state.no_dire_beast_until > now) then
		local db_min = PREC.state.no_dire_beast_until - now;
		if (t < db_min) then
			t = db_min;
		end		
	end

	return {
		t = t,
		waitmana = waitmana,
	};
end

function PREC.abilities.arcane_sv.func(t, now, waitmana)

	-- perform this check for every ability listed below that
	-- is coming up and is higher in the priority list than
	-- this ability (easy because PREC.current_shots is built
	-- from highest priority upwards)
	--
	-- queue up arcane shot only if:
	-- 	current focus
	-- 	+ cost of arcane shot
	-- 	+ regen until high-priority-shot cooldown
	-- is more than:
	-- 	high-priority-shot cost


	--
	-- calculate arcane shot cost/cooldown
	--

	local _, _, _, shot_cost = GetSpellInfo("Arcane Shot");
	local shot_start, shot_dur = GetSpellCooldown("Arcane Shot");
	local regen_base, regen_cast = GetPowerRegen();

	local now = GetTime();
	local cur_focus = UnitPower("player");

	local shot_cooldown = 0;
	if (shot_start > 0 and shot_dur > 0) then
		shot_cooldown = shot_start + shot_dur - now;
	end


	--
	-- create a map of shots ahead of this one in the rotation
	-- (if they are marked as .ok, which filters out boss shots, etc)
	--

	local higher_shots = {};
	local i;
	local v;

	for i,v in pairs(PREC.current_shots) do

		local ability = PREC.abilities[i];
		if (ability and v.ok) then
			higher_shots[ability.spell] = v;
		end
	end


	--
	-- perform wait/focus check for each high-priority shot
	--

	local reserve_shots = {
		"Explosive Shot",
		"Black Arrow",
		"A Murder of Crows",
		"Glaive Toss",
		"Serpent Sting"
	};

	local i;
	local v;
	for i,v in ipairs(reserve_shots) do

		if (higher_shots[v]) then

			local _, _, _, ss_cost = GetSpellInfo(v);
			local ss_cooldown = higher_shots[v].t;

			-- if shot is going to happen within the next GCD, 
			-- or within GCD after this shot, ignore this shot
			if (ss_cooldown - shot_cooldown <= 1.5) then

				--PREC.debug_data = v.." is soon";
				return {
					hide_now = true,
				};
			end

			local total_focus = cur_focus + (ss_cooldown * regen_cast) - shot_cost;

			--PREC.debug_data = string.format("%f, %f", regen_cast, total_focus);

			if (total_focus < ss_cost) then

				--PREC.debug_data = v.." needs focus";
				return {
					hide_now = true,
				};
			end
		end
	end

	return {
		t = t,
		waitmana = waitmana,
	};
end

